Add  Applications Widget for Launching Configurable App Shortcuts

JustForFun88 opened the pull request:
This pull request introduces a new Applications widget that displays a user-defined list of application launchers in the UI. Each app entry supports an icon, a label, and executes its configured command on click.

The design of this widget is inspired by the **Applications Widget** of **YASB Reborn**. I personally missed this functionality and aimed to bring a similar experience to `komorebi-bar`.

## Features
- Declarative configuration via ApplicationsConfig and per-app AppConfig.
- Supports display modes: icon-only, text-only, or icon + text.
- Icons can be:
  - loaded from image files (.png, .ico, etc.)
  - extracted from executable files on Windows (e.g. .exe)
- Optional display of the launch command on hover.

## Limitations
Only image-based icons and Windows .exe embedded icons are currently supported. I made the attempt to support glyph icons from Nerd Fonts, but it failed—most likely due to the font not being properly loaded or for another reason. This may be revisited in the future.

## Example Configuration
```json
{
  "Applications": {
    "enable": true,
    "show_launched": true,
    "display": "Icon",
    "applications": [
      {
        "name": "Quick Settings",
        "icon": ".../settings-2.png",
        "command": "explorer.exe ms-actioncenter:controlcenter/true"
      },
      {
        "name": "Task Manager",
        "icon": ".../Windows_Task_Manager_Icon.png",
        "command": "cmd /c Taskmgr"
      },
      {
        "name": "WezTerm",
        "command": "C:\\Program Files\\WezTerm\\wezterm-gui.exe"
      },
      {
        "name": "VS Code",
        "command": "C:\\Program Files\\Microsoft VS Code\\Code.exe"
      }
    ]
  }
}
```

## Full Configuration Schema

Below is the full set of available configuration options for the Applications widget:

**Applications (root)**

| Field          | Type     | Default     | Description                                                                |
|----------------|----------|-------------|----------------------------------------------------------------------------|
| enable         | bool     | required    | Enables or disables the applications widget                                |
| show_launched  | bool     | false       | *(optional)* Show the launch command on hover                              |
| spacing        | f32      | --          | *(optional)* Horizontal spacing (in pixels) between application buttons    |
| display        | string   | "Icon"      | *(optional)* Default **Display Format** for all applications               |
| applications   | array    | required    | List of application entries. Each item must follow `applications` config   |

**applications (per application entry)**

| Field          | Type     | Default                            | Description                                                        |
|----------------|----------|------------------------------------|--------------------------------------------------------------------|
| enable         | bool     | inherited from Applications        | *(optional)* Enables or disables this individual app entry         |
| show_launched  | bool     | inherited from Applications        | *(optional)* Show launch command on hover                          |
| name           | string   | required                           | Display name of the application                                    |
| icon           | string   | extracted from executable (if any) | *(optional)* Path to an image file to use as the icon              |
| command        | string   | required                           | Command to execute (e.g. path to the application or shell command) |
| display        | string   | inherited from Applications        | *(optional)* **Display Format**: "Icon", "Text", or "IconAndText"  |

**Display Format (string values)**

| Value          | Description                          |
|----------------|--------------------------------------|
| "Icon"         | Show only the application icon       |
| "Text"         | Show only the application name       |
| "IconAndText"  | Show both icon and name              |


https://github.com/user-attachments/assets/80077210-8b67-4f4e-bc03-f523821aa5e3

LGUG2Z commented:
Spasibo! This demo looks very cool indeed! I'm a little behind with life stuff at the moment, but maybe @CtByte has time to look at the contents of the PR and make any comments?

CtByte commented:
It looks promising. I don't have much time, but it's a very good start for sure. I do have a few suggestions for later, but it's not something that this must include right now.

I will try to carve out a few minutes to check this out later!

JustForFun88 commented:
> It looks promising at first glance and I really like all the comments! I don't have much time, but it's a very good start for sure. I do have a few suggestions for later, but it's not something that this must include right now.
> 
> I will try to carve out a few minutes to check this out later!
> 
> @JustForFun88 one suggestion for sure is to run the `just jsonschema` command.

`just jsonschema` has been run. Also added a launch cooldown to prevent accidental double-clicks. Let me know if anything else is needed!

CtByte commented:
I do not think this is needed right now, the PR looks good as it is, I just have a few ideas that I would like to share . This could be added later on, or at least you could tell me what you think.

I think it would be nice to have a collapsible section like the layout widget (or much like the widows sys-tray) for apps you don't want to take up space on your bar. It can even be another window, like the start menu, that can open like a popup window.

![image](https://github.com/user-attachments/assets/9d0dfa38-191f-42b4-8556-0d1b65334d80)

---

Also, would it be possible to make "predefined buttons" to prompt the windows shut down window for example?

This opens the "shut down windows popup"
```powershell
(New-Object -ComObject Shell.Application).ShutdownWindows()
``` 

![image](https://github.com/user-attachments/assets/7271a507-dd7d-40fa-9d73-4693e31fa748)

Someone asked for buttons for logout/shut down/restart/lock widget on the bar, perhaps this is the perfect place?
It is possible to run something like this, but I am not sure if it makes sense. It could be a yes-no prompt executing these commands when the user says "yes"

```powersell
Add-Type -AssemblyName System.Windows.Forms
if ( [System.Windows.Forms.MessageBox]::Show('Proceed?','Title',[System.Windows.Forms.MessageBoxButtons]::YesNo) -eq 'Yes' ) {
  'User chose Yes'
} else {
  'User chose No'
}
```

I am thinking it could be a new type of application that is predefined like this
```json
{
  "Applications": {
    "enable": true,
    "show_launched": true,
    "display": "Icon",
    "applications": [
      {
        "type": "ShutdownWindows"
      },
      {
        "type": "LockScreen"
      },
      {
        "type": "ShutDown"
      },
      {
        "name": "Quick Settings",
        "icon": ".../settings-2.png",
        "command": "explorer.exe ms-actioncenter:controlcenter/true"
      },
      // ...
    ]
  }
}
```

Again, these are just ideas, I've had. Great PR btw!

LGUG2Z commented:
I have some requested changes after taking a look at this and getting it running:

`show_launched` should be renamed to `show_command_on_hover` - I initially expected this to show an indicator if the associated process with the command is running based on the current option name, but in the interests of simplifying this functionality can probably be omitted completely

`applications` on the `ApplicationsConfig` struct should also be to `items` to avoid `Applications.applications` in the JSON object

`icon` should take an `Option<PathBuf>` to use the type system to ensure that a valid path is passed, and this should hook use the path extension to resolve any environment variables

And since there are no plans for the bar to be ported to any other platform for now we can remove the os system check cfg block and the fallback to `sh`

LGUG2Z commented:
Also please make use of the process icon cache and make sure to use the `windows_icons_fallback` hack (ref here: https://github.com/LGUG2Z/komorebi/blob/5e308b9131983116f89cf5610821c3dbdfa574d6/komorebi-bar/src/widgets/komorebi.rs#L904)

JustForFun88 commented:
> `icon` should take an `Option<PathBuf>` to use the type system to ensure that a valid path is passed, and this should hook use the path extension to resolve any environment variables

Thanks for the feedback!

Regarding PathBuf — I'm a bit hesitant since not all configured commands could be file paths (e.g. "cmd /c Taskmgr"). OsString might be a safer alternative if we want to reflect that.

As for icon caching — I intentionally skipped it because the icon is already stored as an RgbaImage in the App struct to avoid reloading. But if widgets are recreated often, I’m happy to use the shared cache. I haven’t looked closely at the config loading and parsing code yet, but I assume `komorebi-bar` doesn’t reinitialize widgets unless the configuration file has changed

LGUG2Z commented:
> As for icon caching — I intentionally skipped it because the icon is already stored as an RgbaImage in the App struct to avoid reloading.

This makes sense - in this case just make sure to add the `windows_icon_fallback` hack that I referenced in the code block above

> Regarding PathBuf — I'm a bit hesitant since not all configured commands could be file paths (e.g. "cmd /c Taskmgr"). OsString might be a safer alternative if we want to reflect that.

Just to clarify having a `String` for the `command` is fine - I was referring to the `icon` config option which is currently `Option<String>` but should always refer to a local filepath if specified (unless I'm missing something?)

JustForFun88 commented:
@LGUG2Z Noted and implemented the requested changes:

- `show_launched` has been renamed to `show_command_on_hover` as suggested.
- `applications` on ApplicationsConfig has been renamed to `items` to avoid redundancy.

> > As for icon caching — I intentionally skipped it because the icon is already stored as an RgbaImage in the App struct to avoid reloading.
> 
> This makes sense - in this case just make sure to add the `windows_icon_fallback` hack that I referenced in the code block above

- Icon fallback is already implemented as shown below:
  ```rust
  fn draw_icon(&self, ctx: &Context, ui: &mut Ui, icon_config: &IconConfig) {
      if let Some(icon) = &self.icon {
          icon.draw(ctx, ui, icon_config);
      } else {
          Icon::draw_fallback(ui, Vec2::splat(icon_config.size));
      }
  }
  ```

> > Regarding PathBuf — I'm a bit hesitant since not all configured commands could be file paths (e.g. "cmd /c Taskmgr"). OsString might be a safer alternative if we want to reflect that.
> 
> Just to clarify having a `String` for the `command` is fine - I was referring to the `icon` config option which is currently `Option<String>` but should always refer to a local filepath if specified (unless I'm missing something?)

Regarding `icon: Option<String>`, I originally intended to also support Nerd Font icons, but it didn’t work initially. I later realized my config was missing a proper font definition. I’ve now fixed that and updated the `struct Icon` to `enum Icon` accordingly:
```rust
/// Holds decoded image data to be used as an icon in the UI.
#[derive(Clone, Debug)]
pub enum Icon {
    Image(RgbaImage),
    Text(String),
}
```
This now allows both **image-based** icons and **Nerd Font glyphs**, like:
```json
  "center_widgets": [
    {
      "Applications": {
        "enable": true,
        "show_command_on_hover": true,
        "display": "Icon",
        "items": [
          {
            "name": "Quick Settings",
            "icon": "\ueb51",
            "command": "explorer.exe ms-actioncenter:controlcenter/true"
          },
          {
            "name": "Task Manager",
            "icon": "\uf422",
            "command": "cmd /c Taskmgr"
          },
          {
            "name": "Windows Terminal",
            "icon": "C:/Users/LENOVO/.config/yasb/icon/terminal.ico",
            "command": "wt",
            "show_launched": true
          },
          {
            "name": "WezTerm",
            "command": "C:\\Program Files\\WezTerm\\wezterm-gui.exe"
          }
        ]
      }
    }
  ],
```

CtByte commented:
> my config was missing a proper font definition.

did you use the `font_family` setting on the bar's config? I do not have the heart to change my `JetBrains Mono` font but I also want Nerd Fonts :)

JustForFun88 commented:
> > my config was missing a proper font definition.
> 
> did you use the `font_family` setting on the bar's config? I do not have the heart to change my `JetBrains Mono` font but I also want Nerd Fonts :)
> 
> I would be cool to support that for the workspace names as well 🤔

That was exactly what I missed — once I enabled Nerd Fonts via the `font_family` setting in the bar config, everything worked perfectly!

But enabling it also revealed a small bug in other widgets: they use **egui_phosphor::Variant::Regular**, and it seems the user-defined font overrides it, which results in some strange-looking icons in the other widgets😄 Still, that's something that can definitely be addressed separately.

CtByte commented:
I wonder if just making the `font_family` a Vec and register multiple fonts would be the way to go. There might be some overlap between phosphor and nerd font.

There is a [nerdfont for egui](https://github.com/bernsteining/egui_nerdfonts) also, but the project seems to be inactive (despite my pr efforts :( )

JustForFun88 commented:
> I wonder if just making the `font_family` a Vec and register multiple fonts would be the way to go. Not sure if there is a way to overcome the overlap between phosphor and nerd font.
> 
> There is a [nerdfont for egui](https://github.com/bernsteining/egui_nerdfonts) also, but the project seems to be inactive (despite my pr efforts 😞 )

Maybe it would actually be better to remove the use of **egui_phosphor::Variant::Regular** altogether and switch to using the **SymbolsOnly** font from the Nerd Font project. It would simplify things and also give access to a much broader set of glyphs out of the box.

CtByte commented:
That is up for debate. Originally, the komorebi bar was meant to be a very simple and quick-to-setup bar so you would not end up ricing your bar like yasb or zebar for months/years. 

The option is always nice to have (and the phosphor pack is limited for sure), but just have a look at the amount of hand-picked icons on the time and network widgets alone. It would take a long time to replace every single icon from another icon pack.

JustForFun88 commented:
> That is up for debate. Originally, the komorebi bar was meant to be a very simple and quick-to-setup bar so you would not end up ricing your bar like yasb or zebar for months/years.
> 
> The option is always nice to have (and the phosphor pack is limited for sure), but just have a look at the amount of hand-picked icons on the time and network widgets alone. It would take a long time to replace every single icon from another icon pack.

What I had in mind was not requiring the user to install the `SymbolsOnly` font themselves, but rather embedding it in the same way `egui_phosphor::Variant::Regular` currently works. After all, that variant is just embedding the font bytes:
```
Variant::Regular => &*include_bytes!("../../res/Phosphor.ttf"),
```
The idea would be to apply the same strategy to `SymbolsOnly`, which would enable broader glyph support while keeping the setup seamless for users. Of course, the current set of icons is already very well curated — this would simply offer an additional option for those who need more symbols without compromising the simplicity of the bar.

CtByte commented:
💡 Oh, I was not aware of this, it is, for sure, worth looking into!

LGUG2Z commented:
> Maybe it would actually be better to remove the use of egui_phosphor::Variant::Regular altogether and switch to using the SymbolsOnly font from the Nerd Font project. It would simplify things and also give access to a much broader set of glyphs out of the box.

Let's separate this discussion out from this PR - for now I think having a doc note that says using Nerd glyphs requires using a font_family that supports them is sufficient

LGUG2Z commented on komorebi-bar/src/widgets/applications.rs:
```diff
@@ -0,0 +1,326 @@
+    /// Draws a fallback icon when the specified icon cannot be loaded.
+    /// Displays a simple crossed-out rectangle as a placeholder.
+    #[inline]
+    pub fn draw_fallback(ui: &mut Ui, icon_size: Vec2) {
```
This is not quite the same as the `windows_icons_fallback`

`windows_icons_fallback` reverts to an older variant of the icon lookup algorithm which often succeeds in cases when the `windows_icons` algorithm fails

What you have here seems more like a fallback to that 2nd-level fallback

JustForFun88 replied:
Fixed

LGUG2Z commented on komorebi-bar/src/widgets/applications.rs:
```diff
@@ -0,0 +1,326 @@
+
+        let path = std::path::Path::new(&config.command);
+        if path.is_file() {
+            return windows_icons::get_icon_by_path(&config.command).map(Icon::Image);
```
This is the bit where if this doesn't return an icon, we should return whatever we get from `windows_icons_fallback`, even if that also returns nothing

JustForFun88 replied:
Fixed

LGUG2Z commented:
I think this is okay to merge - I have a few more changes I'll on top once this is on `master` regarding making the icon extraction from exe files a little smarter, and adding automatic environment variable resolution so that it's possible to have commands like `$Env:USERPROFILE/scoop/apps/zed-nightly/current/zed.exe`

LGUG2Z commented:
Merged this just now - Rahmat @JustForFun88!
